Send an explicit interrupt to cancel an attach waitfor.

jimingham opened the pull request:
Currently when you interrupt a:

(lldb) process attach -w -n some_process

lldb just closes the connection to the stub and kills the lldb_private::Process it made for the attach.  The stub at the other end notices the connection go down and exits because of that.   But when communication to a device is handled through some kind of proxy server which isn't as well behaved as one would wish, that signal might not be reliable, causing debugserver to persist on the machine, waiting to steal the next instance of that process.

We can work around those failures by sending an explicit interrupt before closing down the connection.  The stub will also have to be waiting for the interrupt for this to make any difference.  I changed debugserver to do that.

I didn't make the equivalent change in lldb-server.  So long as you aren't faced with a flakey connection, this should not be necessary.

jimingham committed:
Send an explicit interrupt to cancel an attach waitfor.

Currently when you interrupt a:

(lldb) process attach -w -n some_process

lldb just closes the connection to the stub and kills the process it
made for the attach.  The stub at the other end notices the connection
go down and exits because of that.  But when communication to a device
is handled through some kind of proxy server, that signal might not
be reliable, causing debugserver to persist on the machine.

We can work around those failures by sending an explicit interrupt
before closing down the connection.  The stub will also have to be
waiting for the interrupt for this to make any difference.  I changed
debugserver to do that.

I didn't make the equivalent change in lldb-server.  So long as you
aren't faced with a flakey connection, this should not be necessary.

jasonmolenda reviewed (approved):
LGTM.

JDevlieghere commented on lldb/source/Target/Process.cpp:
```diff
@@ -3843,6 +3843,13 @@ thread_result_t Process::RunPrivateStateThread(bool is_secondary_thread) {
+        // and that will get it to shut down.  But there are remote debugging
+        // cases where relying on that side-effect causes the shutdown to be 
+        // flakey, so we should send a positive signal to interrupt the wait. 
+        Status error = HaltPrivate();
```
Should we log the return value?

jimingham replied:
Sure, I'll add that.

JDevlieghere commented on lldb/tools/debugserver/source/DNB.cpp:
```diff
@@ -799,7 +799,33 @@ DNBProcessAttachWait(RNBContext *ctx, const char *waitfor_process_name,
+        // Over here in DNB land we can see that there was a packet, but all
+        // the methods to actually handle it are protected.  It's not worth
+        // rearranging all that just to get which packet we were sent...
+        DNBLogError("Interrupted by packet while waiting for '%s' to appear.\n",
```
Can we check the packet and assert/log when it's not an interrupt? I know that's technically not allowed by the GDB remote protocol, but that would make it easy to catch such a mistake.

jimingham replied:
All the code that does the actual packet handling (as opposed to just waiting for a packet type) is private to RNBRemote.cpp, and not available here.  I didn't think getting the packet here was worth rejiggering that relationship.  You can turn on packet logging in debugserver if you wanted to see which packet this was.

jasonmolenda replied:
The structure of the gdb remote protocol is such that when you've sent a packet, you can't send anything else until you get a response.  "continue" is special in that you can send ^C to the remote stub to interrupt execution.  A "waitfor" attach request most reasonably behaves the same as "continue" - the only thing it would make any sense for the debugger to send when we're waiting is ^C.  If the debugger were to send anything else, yeah, something has gone very wrong, but when Jim described the layering and how it wasn't straightforward to check that, I didn't push further, it seemed fine to me to assume the debugger didn't do something unsupported like send a non-interrupt packet.

clayborg replied:
I agree that we are attaching and the only thing we can send a ^C, so no need for anything else here really.

clayborg reviewed (commented):
Looks good to me from a code perspective as long as this works.

It would be nice to add a test for Darwin only. Luckily this is easy to test to ensure it works.

jimingham committed:
Add a test case for cancelling attach.  This should work everywhere,
since cancelling attach actually does work everywhere that we are
currently testing.

jimingham commented:
I added a test case for cancelling attach.  This should actually work everywhere so I didn't limit it to Darwin.  It doesn't test the actual bug I was trying to fix because that would require coming up with an unreliable file handle repeater and that's way more work than it's worth.  But since this uses Python threading, and interruption so I wouldn't be surprised if it doesn't work reliably everywhere.
